### ISSUE TITLE
PlotlyJS not saving to HTML properly

### ISSUE THREAD

**Issue by asinghvi17** [JuliaPlots/Plots.jl#1934]
As discussed on Slack by @mkborregaard and I, the `plotlyjs` and `plotly` backends both have issues saving to HTML.  As I see it, there are two main tasks to accomplish here:

- [ ] Fix random omission of `<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>` from `plotly`

- [x] figure out what `plotlyjs` is doing, suspicion is that show(io, MIME"text/html, p) actually shows what is passed to Blink.jl or Electron.

**Comment by asinghvi17** [comment]
`plotly` output:

```html
<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>    <div id="7a9734c4-ba1a-457d-8162-0c254bc4086d" style="width:600px;height:400px;"></div>
    <script>
    PLOT = document.getElementById('7a9734c4-ba1a-457d-8162-0c254bc4086d');
    Plotly.plot(PLOT, [{"xaxis":"x1","colorbar":{"title":""},"yaxis":"y1","text":[null,null],"x":[0,1],"showlegend":true,"mode":"lines","name":"y1","zmin":-0.1,"legendgroup":"y1","z":[0,1],"zmax":0.1,"line":{"color":"rgba(0, 154, 250, 1.000)","shape":"linear","dash":"solid","width":1},"y":[0,1],"type":"scatter3d","hoverinfo":"text"}], {"showlegend":true,"paper_bgcolor":"rgba(255, 255, 255, 1.000)","annotations":[],"height":400,"margin":{"l":0,"b":20,"r":0,"t":20},"plot_bgcolor":"rgba(255, 255, 255, 1.000)","scene":{"camera":{"eye":{"y":-1.9499999999999997,"z":1.3,"x":1.12583302491977}},"xaxis":{"showticklabels":true,"gridwidth":0.5,"tickvals":[0.0,0.25,0.5,0.75,1.0],"visible":true,"ticks":"inside","range":[-0.03,1.03],"domain":[0.0658209390492855,0.9934383202099738],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["0.00","0.25","0.50","0.75","1.00"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"y1"},"yaxis":{"showticklabels":true,"gridwidth":0.5,"tickvals":[0.0,0.25,0.5,0.75,1.0],"visible":true,"ticks":"inside","range":[-0.03,1.03],"domain":[0.03762029746281716,0.9901574803149606],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["0.00","0.25","0.50","0.75","1.00"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)","anchor":"x1"},"zaxis":{"showticklabels":true,"gridwidth":0.5,"tickvals":[0.0,0.25,0.5,0.75,1.0],"visible":true,"ticks":"inside","range":[-0.03,1.03],"tickmode":"array","linecolor":"rgba(0, 0, 0, 1.000)","showgrid":true,"title":"","mirror":false,"tickangle":0,"showline":true,"gridcolor":"rgba(0, 0, 0, 0.100)","titlefont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":15},"tickcolor":"rgb(0, 0, 0)","ticktext":["0.00","0.25","0.50","0.75","1.00"],"zeroline":false,"type":"-","tickfont":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"zerolinecolor":"rgba(0, 0, 0, 1.000)"}},"legend":{"tracegroupgap":0,"bordercolor":"rgba(0, 0, 0, 1.000)","bgcolor":"rgba(255, 255, 255, 1.000)","font":{"color":"rgba(0, 0, 0, 1.000)","family":"sans-serif","size":11},"y":1.0,"x":1.0},"width":600});
    </script>
```

**Comment by asinghvi17** [comment]
`plotlyjs` output:

```html
<div class='display:none'></div><unsafe-script style='display:none'>
WebIO.mount(this.previousSibling,{&quot;props&quot;:{},&quot;nodeType&quot;:&quot;Scope&quot;,&quot;type&quot;:&quot;node&quot;,&quot;instanceArgs&quot;:{&quot;imports&quot;:{&quot;data&quot;:[{&quot;name&quot;:&quot;Plotly&quot;,&quot;type&quot;:&quot;js&quot;,&quot;url&quot;:&quot;/assetserver/3ac5cd54f2013e7e4e31d2d072f4bde7b7f8c3e1-plotly-latest.min.js&quot;},{&quot;name&quot;:null,&quot;type&quot;:&quot;js&quot;,&quot;url&quot;:&quot;/assetserver/dac9edc9356c92574fe5d72731f2b28829761d91-plotly_webio.bundle.js&quot;}],&quot;type&quot;:&quot;async_block&quot;},&quot;id&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;handlers&quot;:{&quot;_toImage&quot;:[(function (options){return this.Plotly.toImage(this.plotElem,options).then((function (data){return WebIO.setval({&quot;name&quot;:&quot;image&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_42&quot;,&quot;type&quot;:&quot;observable&quot;},data)}))})],&quot;__get_gd_contents&quot;:[(function (prop){prop==&quot;data&quot; ? (WebIO.setval({&quot;name&quot;:&quot;__gd_contents&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_43&quot;,&quot;type&quot;:&quot;observable&quot;},this.plotElem.data)) : undefined; return prop==&quot;layout&quot; ? (WebIO.setval({&quot;name&quot;:&quot;__gd_contents&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_43&quot;,&quot;type&quot;:&quot;observable&quot;},this.plotElem.layout)) : undefined})],&quot;_promises&quot;:{&quot;importsLoaded&quot;:[(function (Plotly){var gd=this.dom.querySelector(&quot;#plot-13002901-1d3c-40cc-88a5-2ebeaeace49d&quot;); this.plotElem=gd; this.Plotly=Plotly; (window.Blink!==undefined) ? (gd.style.width=&quot;100%&quot;, gd.style.height=&quot;100vh&quot;, gd.style.marginLeft=&quot;0%&quot;, gd.style.marginTop=&quot;0vh&quot;) : undefined; window.onresize=(function (){return Plotly.Plots.resize(gd)}); Plotly.newPlot(gd,[],{&quot;margin&quot;:{&quot;l&quot;:50,&quot;b&quot;:50,&quot;r&quot;:50,&quot;t&quot;:60}},{}); gd.on(&quot;plotly_hover&quot;,(function (data){var filtered_data=WebIO.PlotlyCommands.filterEventData(gd,data,&quot;hover&quot;); return !(filtered_data.isnil) ? (WebIO.setval({&quot;name&quot;:&quot;hover&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_38&quot;,&quot;type&quot;:&quot;observable&quot;},filtered_data.out)) : undefined})); gd.on(&quot;plotly_unhover&quot;,(function (){return WebIO.setval({&quot;name&quot;:&quot;hover&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_38&quot;,&quot;type&quot;:&quot;observable&quot;},{})})); gd.on(&quot;plotly_selected&quot;,(function (data){var filtered_data=WebIO.PlotlyCommands.filterEventData(gd,data,&quot;selected&quot;); return !(filtered_data.isnil) ? (WebIO.setval({&quot;name&quot;:&quot;selected&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_39&quot;,&quot;type&quot;:&quot;observable&quot;},filtered_data.out)) : undefined})); gd.on(&quot;plotly_deselect&quot;,(function (){return WebIO.setval({&quot;name&quot;:&quot;selected&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_39&quot;,&quot;type&quot;:&quot;observable&quot;},{})})); gd.on(&quot;plotly_relayout&quot;,(function (data){var filtered_data=WebIO.PlotlyCommands.filterEventData(gd,data,&quot;relayout&quot;); return !(filtered_data.isnil) ? (WebIO.setval({&quot;name&quot;:&quot;relayout&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_41&quot;,&quot;type&quot;:&quot;observable&quot;},filtered_data.out)) : undefined})); return gd.on(&quot;plotly_click&quot;,(function (data){var filtered_data=WebIO.PlotlyCommands.filterEventData(gd,data,&quot;click&quot;); return !(filtered_data.isnil) ? (WebIO.setval({&quot;name&quot;:&quot;click&quot;,&quot;scope&quot;:&quot;scope-4da27e6d-6a96-4d35-88a1-f84224b63b99&quot;,&quot;id&quot;:&quot;ob_40&quot;,&quot;type&quot;:&quot;observable&quot;},filtered_data.out)) : undefined}))})]},&quot;_downloadImage&quot;:[(function (options){return this.Plotly.downloadImage(this.plotElem,options)})],&quot;_commands&quot;:[(function (args){var fn=args.shift(); var elem=this.plotElem; var Plotly=this.Plotly; args.unshift(elem); return Plotly[fn].apply(this,args)})]},&quot;systemjs_options&quot;:null,&quot;observables&quot;:{&quot;_toImage&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_45&quot;,&quot;value&quot;:{}},&quot;hover&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_38&quot;,&quot;value&quot;:{}},&quot;selected&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_39&quot;,&quot;value&quot;:{}},&quot;__gd_contents&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_43&quot;,&quot;value&quot;:{}},&quot;click&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_40&quot;,&quot;value&quot;:{}},&quot;image&quot;:{&quot;sync&quot;:true,&quot;id&quot;:&quot;ob_42&quot;,&quot;value&quot;:&quot;&quot;},&quot;__get_gd_contents&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_47&quot;,&quot;value&quot;:&quot;&quot;},&quot;_downloadImage&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_46&quot;,&quot;value&quot;:{}},&quot;relayout&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_41&quot;,&quot;value&quot;:{}},&quot;_commands&quot;:{&quot;sync&quot;:false,&quot;id&quot;:&quot;ob_44&quot;,&quot;value&quot;:[&quot;relayout&quot;,{&quot;showlegend&quot;:true,&quot;paper_bgcolor&quot;:&quot;rgba(255, 255, 255, 1.000)&quot;,&quot;annotations&quot;:[],&quot;height&quot;:400,&quot;margin&quot;:{&quot;l&quot;:0,&quot;b&quot;:20,&quot;r&quot;:0,&quot;t&quot;:20},&quot;plot_bgcolor&quot;:&quot;rgba(255, 255, 255, 1.000)&quot;,&quot;scene&quot;:{&quot;camera&quot;:{&quot;eye&quot;:{&quot;y&quot;:-1.9499999999999997,&quot;z&quot;:1.3,&quot;x&quot;:1.12583302491977}},&quot;xaxis&quot;:{&quot;showticklabels&quot;:true,&quot;gridwidth&quot;:0.5,&quot;tickvals&quot;:[0.0,0.25,0.5,0.75,1.0],&quot;visible&quot;:true,&quot;ticks&quot;:&quot;inside&quot;,&quot;range&quot;:[-0.03,1.03],&quot;domain&quot;:[0.0658209390492855,0.9934383202099738],&quot;tickmode&quot;:&quot;array&quot;,&quot;linecolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;showgrid&quot;:true,&quot;title&quot;:&quot;&quot;,&quot;mirror&quot;:false,&quot;tickangle&quot;:0,&quot;showline&quot;:true,&quot;gridcolor&quot;:&quot;rgba(0, 0, 0, 0.100)&quot;,&quot;titlefont&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:15},&quot;tickcolor&quot;:&quot;rgb(0, 0, 0)&quot;,&quot;ticktext&quot;:[&quot;0.00&quot;,&quot;0.25&quot;,&quot;0.50&quot;,&quot;0.75&quot;,&quot;1.00&quot;],&quot;zeroline&quot;:false,&quot;type&quot;:&quot;-&quot;,&quot;tickfont&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:11},&quot;zerolinecolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;anchor&quot;:&quot;y1&quot;},&quot;yaxis&quot;:{&quot;showticklabels&quot;:true,&quot;gridwidth&quot;:0.5,&quot;tickvals&quot;:[0.0,0.25,0.5,0.75,1.0],&quot;visible&quot;:true,&quot;ticks&quot;:&quot;inside&quot;,&quot;range&quot;:[-0.03,1.03],&quot;domain&quot;:[0.03762029746281716,0.9901574803149606],&quot;tickmode&quot;:&quot;array&quot;,&quot;linecolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;showgrid&quot;:true,&quot;title&quot;:&quot;&quot;,&quot;mirror&quot;:false,&quot;tickangle&quot;:0,&quot;showline&quot;:true,&quot;gridcolor&quot;:&quot;rgba(0, 0, 0, 0.100)&quot;,&quot;titlefont&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:15},&quot;tickcolor&quot;:&quot;rgb(0, 0, 0)&quot;,&quot;ticktext&quot;:[&quot;0.00&quot;,&quot;0.25&quot;,&quot;0.50&quot;,&quot;0.75&quot;,&quot;1.00&quot;],&quot;zeroline&quot;:false,&quot;type&quot;:&quot;-&quot;,&quot;tickfont&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:11},&quot;zerolinecolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;anchor&quot;:&quot;x1&quot;},&quot;zaxis&quot;:{&quot;showticklabels&quot;:true,&quot;gridwidth&quot;:0.5,&quot;tickvals&quot;:[0.0,0.25,0.5,0.75,1.0],&quot;visible&quot;:true,&quot;ticks&quot;:&quot;inside&quot;,&quot;range&quot;:[-0.03,1.03],&quot;tickmode&quot;:&quot;array&quot;,&quot;linecolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;showgrid&quot;:true,&quot;title&quot;:&quot;&quot;,&quot;mirror&quot;:false,&quot;tickangle&quot;:0,&quot;showline&quot;:true,&quot;gridcolor&quot;:&quot;rgba(0, 0, 0, 0.100)&quot;,&quot;titlefont&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:15},&quot;tickcolor&quot;:&quot;rgb(0, 0, 0)&quot;,&quot;ticktext&quot;:[&quot;0.00&quot;,&quot;0.25&quot;,&quot;0.50&quot;,&quot;0.75&quot;,&quot;1.00&quot;],&quot;zeroline&quot;:false,&quot;type&quot;:&quot;-&quot;,&quot;tickfont&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:11},&quot;zerolinecolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;}},&quot;legend&quot;:{&quot;tracegroupgap&quot;:0,&quot;bordercolor&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;bgcolor&quot;:&quot;rgba(255, 255, 255, 1.000)&quot;,&quot;font&quot;:{&quot;color&quot;:&quot;rgba(0, 0, 0, 1.000)&quot;,&quot;family&quot;:&quot;sans-serif&quot;,&quot;size&quot;:11},&quot;y&quot;:1.0,&quot;x&quot;:1.0},&quot;width&quot;:600}]}}},&quot;children&quot;:[{&quot;props&quot;:{&quot;id&quot;:&quot;plot-13002901-1d3c-40cc-88a5-2ebeaeace49d&quot;},&quot;nodeType&quot;:&quot;DOM&quot;,&quot;type&quot;:&quot;node&quot;,&quot;instanceArgs&quot;:{&quot;namespace&quot;:&quot;html&quot;,&quot;tag&quot;:&quot;div&quot;},&quot;children&quot;:[]}]})</unsafe-script>
```

**Comment by asinghvi17** [comment]
test script:
```julia
using Plots
plotly()
p = plot(0:1, 0:1, 0:1, fmt=:html)
savefig(p, "test_p.html")
plotlyjs()
p = plot(0:1, 0:1, 0:1, fmt=:html)
savefig(p, "test_j.html")
```

**Comment by asinghvi17** [comment]
How do you get the plot object from the backend?  I was thinking you could send a plot with a PlotlyJS backend to the Plotly html output method, and thereby generate portable HTML plots. 

```julia
julia> plotly()
Plots.PlotlyBackend()

julia> p = scatter(rand(10))

julia> p.o

julia> typeof(p)
Plots.Plot{Plots.PlotlyBackend}

julia> typeof(p.o)
Nothing

julia> backend_object(p)

julia> typeof(ans)
Nothing
```

**Comment by mkborregaard** [comment]
It's not supposed to work like that, unfortunately.

**Comment by ferrolho** [comment]
Works on `(v1.1) pkg> add Plots@0.23.1`. It bothers me that I always have to downgrade whenever I want to export the `html` so I can embed it somewhere else...

Also reporting the same problem: #1985

**Comment by asinghvi17** [comment]
This issue was talking about Plotly randomly omitting the line `<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>` when saving.  It might be that support for Plotly saving to HTML was removed?  I'll give it a look...

**Comment by daschw** [comment]
Should work now for Plotly with #2125.

**Comment by daschw** [comment]
Hopefully closed in #2131
